### ISSUE TITLE
TokenReview audience support for nodes

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#66311]
Once https://github.com/kubernetes/kubernetes/pull/62802 merges, we should add support for audiences to the Kubelet's token review.

This is important as part of the node isolation efforts, as the node could otherwise use tokens sent with requests to the nodes endpoints (e.g. stats requests) to send requests to other nodes.

I think the node should accept 3 types of audiences:

- legacy (no audience) - required for backwards compatibility
- `system:nodes` - token used for all nodes
- `system:node:<nodeName>` - per-node token, preferred

/kind feature
/sig auth
/sig node
/priority important-longterm

**Comment by tallclair** [comment]
/remove-lifecycle rotten

I was thinking about the use case for this a bit more. I think workloads that need to talk to nodes mostly fall into 2 categories:

1. Cluster-level tools, such as monitoring, which need to talk to every node in the cluster
2. Node-level tools (DaemonSets) that only need to talk to the node they're running on

I'm not sure per-node audiences are the right solution for use case 1. In a large cluster, that would mean the worker would need to collect and use a separate token for every node, which would be unwieldy at best. A certificate-based solution seems more practical for this case.

For 2, something with token audiences makes more sense, but still adds a bit of complexity since the worker needs to figure out which node it's running on first. Another idea is a "localhost" audience, which would only be valid if the request comes in over localhost.

**Comment by eloyekunle** [comment]
I will start working on this issue.

**Comment by krmayankk** [comment]
@eloyekunle  are you working on this already ?

**Comment by eloyekunle** [comment]
@krmayankk I didn't get to work on this issue.

**Comment by krmayankk** [comment]
@tallclair @liggitt @mikedanese  is this still relevant ? Can i start looking into this ?

**Comment by tallclair** [comment]
Definitely still relevant, though somewhat nuanced to use correctly. Unless @mikedanese has started working on it, I'd be happy to have you look into it.

**Comment by mikedanese** [comment]
@krmayankk yes please, go ahead.

**Comment by ehashman** [comment]
@tallclair where are we with this? This issue has been frozen and non-updated for ~2y. Was this implemented?

**Comment by tallclair** [comment]
No. I'm not sure we actually want this. It seems like a fairly obvious thing to have in theory, but in practice most services that talk to nodes need to talk to all nodes, and I'm not sure we want to ask them to maintain N separate tokens (clients?) to talk to N nodes. I think client certs might be a better solution for these cases.

@mikedanese WDYT?

**Comment by anguslees** [comment]
Is there a version of the future where we can have node isolation _and_ kubelet authn using node-generic bearer tokens?

I don't think that's possible?  I think we either need node-specific audiences, or we need to deprecate kubelet authn by tokens entirely (remove `--authentication-token-webhook`).   (Or abandon node isolation as a goal)

**Comment by tallclair** [comment]
That's a good framing @anguslees. The Kubelet API needs a lot of work, and I think it's important that we get this authentication story ironed out before iterating on the other pieces (authorization).

The one case where I think node-scoped bearer tokens could make senes is when you have a DaemonSet pod that wants to talk to the Kubelet of the same node it's running on. Although, that might be a problem better solved with an authorization solution. With bound service account tokens, it would probably make more sense to solve this with an authorization solution, which comes back to those broader Kubelet API changes I mentioned.

In conclusion, we really need someone to think through the end-to-end use cases for the Kubelet API to come up with the right solutions.

**Comment by lmktfy** [comment]
also
/sig security
